File: book/petrol-pump/java/chapter4/section4/LifeCycle.java
package chapter4.section4;

import pump.*;
import nz.sodium.*;
import java.util.Optional;

public class LifeCycle {
    public final Stream<Fuel> sStart;
    public final Cell<Optional<Fuel>> fillActive;
    public final Stream<End> sEnd;

    public enum End { END }

    private static Stream<Fuel> whenLifted(Stream<UpDown> sNozzle,
                                          Fuel nozzleFuel) {
        return sNozzle.filter(u -> u == UpDown.UP)
                      .map(u -> nozzleFuel);
    }

    private static Stream<End> whenSetDown(Stream<UpDown> sNozzle,
                Fuel nozzleFuel,
                Cell<Optional<Fuel>> fillActive) {
        return Stream.<End>filterOptional(
            sNozzle.snapshot(fillActive,
                (u,f) -> u == UpDown.DOWN &&
                         f.equals(Optional.of(nozzleFuel))
                                       ? Optional.of(End.END)
                                       : Optional.empty()));
    }

    public LifeCycle(Stream<UpDown> sNozzle1,
                     Stream<UpDown> sNozzle2,
                     Stream<UpDown> sNozzle3) {
        Stream<Fuel> sLiftNozzle =
            whenLifted(sNozzle1, Fuel.ONE).orElse(
            whenLifted(sNozzle2, Fuel.TWO).orElse(
            whenLifted(sNozzle3, Fuel.THREE)));
        CellLoop<Optional<Fuel>> fillActive = new CellLoop<>();
        this.fillActive = fillActive;
        this.sStart = Stream.filterOptional(
            sLiftNozzle.snapshot(fillActive, (newFuel, fillActive_) ->
                fillActive_.isPresent() ? Optional.empty()
                                        : Optional.of(newFuel)));
        this.sEnd = whenSetDown(sNozzle1, Fuel.ONE, fillActive).orElse(
                    whenSetDown(sNozzle2, Fuel.TWO, fillActive).orElse(
                    whenSetDown(sNozzle3, Fuel.THREE, fillActive)));
        fillActive.loop(
            sEnd.map(e -> Optional.<Fuel>empty())
                .orElse(sStart.map(f -> Optional.of(f)))
                .hold(Optional.empty())
        );
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section4/LifeCycle.cs
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section4
{
    internal class LifeCycle
    {
        private static Stream<Fuel> WhenLifted(Stream<UpDown> sNozzle, Fuel nozzleFuel)
        {
            return sNozzle.Filter(u => u == UpDown.Up).Map(u => nozzleFuel);
        }

        private static Stream<End> WhenSetDown(Stream<UpDown> sNozzle, Fuel nozzleFuel, Cell<Maybe<Fuel>> fillActive) =>
            sNozzle.Snapshot(fillActive, (u, f) => u == UpDown.Down && f.Equals(Maybe.Some(nozzleFuel)) ? Maybe.Some(End.Value) : Maybe.None).FilterMaybe();

        public LifeCycle(Stream<UpDown> sNozzle1, Stream<UpDown> sNozzle2, Stream<UpDown> sNozzle3)
        {
            Stream<Fuel> sLiftNozzle =
                WhenLifted(sNozzle1, Fuel.One).OrElse(
                    WhenLifted(sNozzle2, Fuel.Two).OrElse(
                        WhenLifted(sNozzle3, Fuel.Three)));
            CellLoop<Maybe<Fuel>> fillActive = new CellLoop<Maybe<Fuel>>();
            this.FillActive = fillActive;
            this.SStart = sLiftNozzle.Snapshot(fillActive, (newFuel, fillActiveLocal) => fillActiveLocal.Match(_ => Maybe.None, () => Maybe.Some(newFuel))).FilterMaybe();
            this.SEnd = WhenSetDown(sNozzle1, Fuel.One, fillActive).OrElse(
                WhenSetDown(sNozzle2, Fuel.Two, fillActive).OrElse(
                    WhenSetDown(sNozzle3, Fuel.Three, fillActive)));
            fillActive.Loop(
                this.SEnd.Map(e => Maybe<Fuel>.None)
                    .OrElse(this.SStart.Map(Maybe.Some))
                    .Hold(Maybe.None));
        }

        public Cell<Maybe<Fuel>> FillActive { get; }
        public Stream<Fuel> SStart { get; }
        public Stream<End> SEnd { get; }
    }
}